// @amplitude/node/dist/src/constants.js
Object.defineProperty(exports, "__esModule", { value: true });
exports.DEFAULT_OPTIONS = exports.REQUEST_TIMEOUT_MILLIS_DEFAULT = exports.BASE_RETRY_TIMEOUT_DEPRECATED_WARNING = exports.BASE_RETRY_TIMEOUT_DEPRECATED = exports.AMPLITUDE_SERVER_URL = exports.SDK_NAME = void 0;
var types_1 = require("@amplitude/types");
var package_json_1 = require("../package.json");
Object.defineProperty(exports, "SDK_VERSION", { enumerable: true, get: function () { return package_json_1.version; } });
exports.SDK_NAME = 'amplitude-node';
exports.AMPLITUDE_SERVER_URL = 'https://api2.amplitude.com/2/httpapi';
exports.BASE_RETRY_TIMEOUT_DEPRECATED = 100;
// Default 10 second event timeout
exports.BASE_RETRY_TIMEOUT_DEPRECATED_WARNING = 'DEPRECATED. Please use retryTimeouts. It will be converted to retryTimeouts with exponential wait times (i.e. 100ms -> 200ms -> 400ms -> ...)';
exports.REQUEST_TIMEOUT_MILLIS_DEFAULT = 10000;
// The overridable constants of the node SDK
exports.DEFAULT_OPTIONS = {
    serverUrl: exports.AMPLITUDE_SERVER_URL,
    debug: false,
    // 2kb is a safe estimate for a medium size event object. This keeps the SDK's memory footprint roughly
    // under 32 MB.
    maxCachedEvents: 16000,
    logLevel: types_1.LogLevel.None,
    optOut: false,
    retryTimeouts: [100, 100, 200, 200, 400, 400, 800, 800, 1600, 1600, 3200, 3200],
    // The client will instantiate the retry/transport classes if not defined
    retryClass: null,
    transportClass: null,
    // By default, events flush on the next event loop
    uploadIntervalInSec: 0,
    minIdLength: null,
    onRetry: null,
    requestTimeoutMillis: exports.REQUEST_TIMEOUT_MILLIS_DEFAULT,
};
